wxWidgets-2.9.1/src/tiff/man/tiffset.1
.\" $Id: tiffset.1,v 1.2 2005/11/02 11:07:19 dron Exp $
.\"
.\" Copyright (c) 1988-1997 Sam Leffler
.\" Copyright (c) 1991-1997 Silicon Graphics, Inc.
.\"
.\" Permission to use, copy, modify, distribute, and sell this software and 
.\" its documentation for any purpose is hereby granted without fee, provided
.\" that (i) the above copyright notices and this permission notice appear in
.\" all copies of the software and related documentation, and (ii) the names of
.\" Sam Leffler and Silicon Graphics may not be used in any advertising or
.\" publicity relating to the software without the specific, prior written
.\" permission of Sam Leffler and Silicon Graphics.
.\" 
.\" THE SOFTWARE IS PROVIDED "AS-IS" AND WITHOUT WARRANTY OF ANY KIND, 
.\" EXPRESS, IMPLIED OR OTHERWISE, INCLUDING WITHOUT LIMITATION, ANY 
.\" WARRANTY OF MERCHANTABILITY OR FITNESS FOR A PARTICULAR PURPOSE.  
.\" 
.\" IN NO EVENT SHALL SAM LEFFLER OR SILICON GRAPHICS BE LIABLE FOR
.\" ANY SPECIAL, INCIDENTAL, INDIRECT OR CONSEQUENTIAL DAMAGES OF ANY KIND,
.\" OR ANY DAMAGES WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS,
.\" WHETHER OR NOT ADVISED OF THE POSSIBILITY OF DAMAGE, AND ON ANY THEORY OF 
.\" LIABILITY, ARISING OUT OF OR IN CONNECTION WITH THE USE OR PERFORMANCE 
.\" OF THIS SOFTWARE.
.\"
.if n .po 0
.TH TIFFSET 1 "November 21, 2004" "libtiff"
.SH NAME
tiffset \- set a field in a
.SM TIFF
header
.SH SYNOPSIS
.B tiffset
[
options
]
.I filename.tif
.SH DESCRIPTION
.I Tiffset
sets the value of a
.SM TIFF
header to a specified value.
.SH OPTIONS
.TP
.B \-s tagnumber [count] value ...
Set the value of the named tag to the value or values specified.
.TP
.B \-sf tagnumber filename
Set the value of the tag to the contents of filename.  This option is
supported for ASCII tags only.
.SH EXAMPLES
The following example sets the image description tag (270) of a.tif to
the contents of the file descrip:
.RS
.nf
tiffset -sf 270 descrip a.tif
.fi
.RE
.PP
The following example sets the artist tag (315) of a.tif to the string
"Anonymous":
.RS
.nf
tiffset -s 305 Anonymous a.tif
.fi
.RE
.PP
This example sets the resolution of the file a.tif to 300 dpi:
.RS
.nf
tiffset -s 296 2 a.tif
tiffset -s 282 300.0 a.tif
tiffset -s 283 300.0 a.tif
.fi
.RE
.SH "SEE ALSO"
.BR tiffdump (1),
.BR tiffinfo (1),
.BR tiffcp (1),
.BR libtiff (3TIFF)
.PP
Libtiff library home page:
.BR http://www.remotesensing.org/libtiff/


ChangeLog.txt
2009-09-23 14:13  zach.mullen

	* Tests/CMakeLists.txt: Uncommented the dependency of
	  CTestTestNoExe on CTestTestNoBuild so that it will work in
	  parallel now.

2009-09-23 14:10  king

	* Source/cmGlobalXCodeGenerator.cxx: Add Xcode SYMROOT setting for
	  custom targets

	  Xcode 1.5 writes helper scripts at the projectDirPath location
	  for targets that do not set SYMROOT.	We now add SYMROOT to
	  custom targets so that all targets set it.  This prevents Xcode
	  1.5 from touching the source directory now that we always set
	  projectDirPath.

	  See issue #8481.

2009-09-23 14:07  zach.mullen

	* Tests/CTestTestParallel/lockFile.c: Make portable c for Parallel
	  test

2009-09-23 14:01  alex

	* Source/: cmDepends.cxx, cmDepends.h, cmDependsC.cxx,
	  cmDependsC.h, cmDependsJava.cxx, cmDependsJava.h,
	  cmLocalUnixMakefileGenerator3.cxx,
	  cmLocalUnixMakefileGenerator3.h: Major optimization of C/C++
	  dependency scanning.

	  Now only the dependencies for the file where the dependencies
	  actually may have changed are rescanned, before that this was
	  done for all source files even if only one source file had
	  changed.  This reduces e.g. on my machine the time for scanning
	  the dependencies of kdelibs/khtml/ when only one file
	  (khtml_global.cpp) has changed from around 7.5 seconds to 1.2
	  seconds.

	  The tests succeed, it does what I expected it to do on kdelibs,
	  and Brad also reviewed the patch, so I think it should be ok.

	  Alex

2009-09-23 13:09  clinton

	* Source/QtDialog/: CMakeSetupDialog.cxx, QCMakeCacheView.cxx,
	  QCMakeCacheView.h: fix compile warnings

2009-09-23 12:46  hoffman

	* Utilities/Release/: magrathea_release.cmake, release_cmake.cmake:
	  Handle older cvs clients that do not allow for the password to be
	  in the CVSROOT.

2009-09-23 12:45  hoffman

	* Tests/CMakeLists.txt: Add nightly builds for linux windows and
	  mac.

2009-09-23 11:38  zach.mullen

	* Tests/: CMakeLists.txt, CTestTestParallel/CMakeLists.txt,
	  CTestTestParallel/lockFile.c, CTestTestParallel/lockFile.cxx: Set
	  new ctest tests to always run, whether CTEST_TEST_CTEST is
	  enabled or not.  Changed parallel test to be portable.

2009-09-23 10:45  king

	* Source/kwsys/SystemTools.cxx: Fix KWSys SystemTools build on
	  cygwin with -mwin32

	  Commit "Optimize KWSys SystemTools::FileExists on Windows"
	  accidentally added "#undef _WIN32" when including <windows.h> on
	  cygwin, which breaks builds using the -mwin32 flag.  This commit
	  removes that line and fixes the real error it was intended to
	  avoid.

2009-09-23 09:00  zach.mullen

	* Tests/CTestTestParallel/test.cmake.in: CTestTestParallel now
	  submits to public dashboard for easier debugging

2009-09-23 08:48  king

	* Source/cmDocumentVariables.cxx,
	  Source/cmGlobalXCodeGenerator.cxx,
	  Source/cmGlobalXCodeGenerator.h,
	  Tests/SystemInformation/SystemInformation.in: Teach Xcode
	  generator to set XCODE_VERSION

	  We set the variable 'XCODE_VERSION' in the CMake language to the
	  Xcode version string (e.g. "3.1.2").	Platform config files may
	  use it later.

2009-09-23 01:01  bigler

	* Modules/FindCUDA.cmake: Updated formatting of documentation plus
	  a little reorganization.

2009-09-23 00:50  bigler

	* Modules/FindCUDA/run_nvcc.cmake: Added a command to make the
	  output directory.  This is to fix the XCode build that uses a
	  different output directory than other systems, and rather than
	  try to match that we'll just make it.

2009-09-23 00:01  kwrobot

	* Source/kwsys/kwsysDateStamp.cmake: KWSys Nightly Date Stamp

2009-09-22 18:44  clinton

	* Modules/FindQt4.cmake: add support for finding
	  qcollectiongenerator executable.  fixes #9248.

2009-09-22 18:29  clinton

	* Source/QtDialog/: CMakeSetup.cxx, CMakeSetupDialog.cxx: fix issue
	  9346.  add binary directory to window title to make it easier to
	  deal with multiple cmake-gui instances

2009-09-22 17:08  hoffman

	* Utilities/Release/dash2win64_release.cmake: new windows build
	  machine for cmake

2009-09-22 17:07  hoffman

	* Utilities/Release/vogon_cygwin.cmake: disable svn

2009-09-22 16:28  clinton

	* Modules/FindQt4.cmake: Add support for Qt configured with custom
	  qtlibinfix (see issue 9571).	Also fix CMP 15 warnings.

2009-09-22 16:18  king

	* Source/: cmGlobalXCodeGenerator.cxx, cmGlobalXCodeGenerator.h:
	  Fix Xcode project references to the source tree

	  Xcode project source file references need to always be relative
	  to the top of the source tree in order for SCM and debug symbols
	  to work right.  We must even allow the relative paths to cross
	  outside of the top source or build directories.

	  For subdirectory project() command Xcode projects we use the
	  source directory containing the project() command as the top.
	  Relative paths are generated accordingly for each subproject.

	  See issue #8481.

2009-09-22 16:16  king

	* Source/: cmLocalGenerator.cxx, cmLocalGenerator.h: Optionally
	  force conversion to relative path

	  In cmLocalGenerator::ConvertToRelativePath we normally convert to
	  relative path only if the local and remote paths both lie inside
	  the source tree or both lie inside the build tree.  This commit
	  adds an optional 'force' argument to allow conversion even when
	  this rule is violated.

2009-09-22 16:12  hoffman

	* bootstrap: Make sure KWSYS_DO_NOT_CLEAN_PUTENV is defined at
	  bootstrap time for cmake in the bootstrap script.

2009-09-22 16:02  king

	* Modules/CMakeDetermineCompilerABI.cmake: Skip implicit link info
	  for multiple OS X archs

	  Implicit link information contains architecture-specific
	  libraries and directories.  The link information cannot be
	  explicitly specified safely when CMAKE_OSX_ARCHITECTURES contains
	  more than one architecture.


wxWidgets-2.9.1/src/tiff/tools/tiffset.c
/******************************************************************************
 * $Id$
 *
 * Project:  libtiff tools
 * Purpose:  Mainline for setting metadata in existing TIFF files.
 * Author:   Frank Warmerdam, warmerdam@pobox.com
 *
 ******************************************************************************
 * Copyright (c) 2000, Frank Warmerdam
 *
 * Permission to use, copy, modify, distribute, and sell this software and 
 * its documentation for any purpose is hereby granted without fee, provided
 * that (i) the above copyright notices and this permission notice appear in
 * all copies of the software and related documentation, and (ii) the names of
 * Sam Leffler and Silicon Graphics may not be used in any advertising or
 * publicity relating to the software without the specific, prior written
 * permission of Sam Leffler and Silicon Graphics.
 * 
 * THE SOFTWARE IS PROVIDED "AS-IS" AND WITHOUT WARRANTY OF ANY KIND, 
 * EXPRESS, IMPLIED OR OTHERWISE, INCLUDING WITHOUT LIMITATION, ANY 
 * WARRANTY OF MERCHANTABILITY OR FITNESS FOR A PARTICULAR PURPOSE.  
 * 
 * IN NO EVENT SHALL SAM LEFFLER OR SILICON GRAPHICS BE LIABLE FOR
 * ANY SPECIAL, INCIDENTAL, INDIRECT OR CONSEQUENTIAL DAMAGES OF ANY KIND,
 * OR ANY DAMAGES WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS,
 * WHETHER OR NOT ADVISED OF THE POSSIBILITY OF DAMAGE, AND ON ANY THEORY OF 
 * LIABILITY, ARISING OUT OF OR IN CONNECTION WITH THE USE OR PERFORMANCE 
 * OF THIS SOFTWARE.
 ******************************************************************************
 *
 * $Log: tiffset.c,v $
 * Revision 1.11  2005/09/13 14:13:42  dron
 * Avoid warnings.
 *
 * Revision 1.10  2005/02/24 14:47:11  fwarmerdam
 * Updated header.
 *
 */


#include <stdio.h>
#include <string.h>
#include <stdlib.h>

#include "tiffio.h"

static char* usageMsg[] = {
"usage: tiffset [options] filename",
"where options are:",
" -s <tagname> [count] <value>...   set the tag value",
" -sf <tagname> <filename>  read the tag value from file (for ASCII tags only)",
NULL
};

static void
usage(void)
{
	int i;
	for (i = 0; usageMsg[i]; i++)
		fprintf(stderr, "%s\n", usageMsg[i]);
	exit(-1);
}

static const TIFFFieldInfo *
GetField(TIFF *tiff, const char *tagname)
{
    const TIFFFieldInfo *fip;

    if( atoi(tagname) > 0 )
        fip = TIFFFieldWithTag(tiff, (ttag_t)atoi(tagname));
    else
        fip = TIFFFieldWithName(tiff, tagname);

    if (!fip) {
        fprintf( stderr, "Field name %s not recognised.\n", tagname );
        return (TIFFFieldInfo *)NULL;
    }

    return fip;
}

int
main(int argc, char* argv[])
{
    TIFF *tiff;
    int  arg_index;

    if (argc < 2)
        usage();

    tiff = TIFFOpen(argv[argc-1], "r+");
    if (tiff == NULL)
        return 2;

    for( arg_index = 1; arg_index < argc-1; arg_index++ ) {
        if (strcmp(argv[arg_index],"-s") == 0 && arg_index < argc-3) {
            const TIFFFieldInfo *fip;
            const char *tagname;

            arg_index++;
            tagname = argv[arg_index];
            fip = GetField(tiff, tagname);

            if (!fip)
                return 3;

            arg_index++;
            if (fip->field_type == TIFF_ASCII) {
                if (TIFFSetField(tiff, fip->field_tag, argv[arg_index]) != 1)
                    fprintf( stderr, "Failed to set %s=%s\n",
                             fip->field_name, argv[arg_index] );
            } else if (fip->field_writecount > 0) {
                int     ret = 1;
                short   wc;

                if (fip->field_writecount == TIFF_VARIABLE)
                        wc = atoi(argv[arg_index++]);
                else
                        wc = fip->field_writecount;

                if (argc - arg_index < wc) {
                    fprintf( stderr,
                             "Number of tag values is not enough. "
                             "Expected %d values for %s tag, got %d\n",
                             wc, fip->field_name, argc - arg_index);
                    return 4;
                }
                    
                if (wc > 1) {
                        int     i, size;
                        void    *array;

                        switch (fip->field_type) {
                                /*
                                 * XXX: We can't use TIFFDataWidth()
                                 * to determine the space needed to store
                                 * the value. For TIFF_RATIONAL values
                                 * TIFFDataWidth() returns 8, but we use 4-byte
                                 * float to represent rationals.
                                 */
                                case TIFF_BYTE:
                                case TIFF_ASCII:
                                case TIFF_SBYTE:
                                case TIFF_UNDEFINED:
				default:
                                    size = 1;
                                    break;

                                case TIFF_SHORT:
                                case TIFF_SSHORT:
                                    size = 2;
                                    break;

                                case TIFF_LONG:
                                case TIFF_SLONG:
                                case TIFF_FLOAT:
                                case TIFF_IFD:
                                case TIFF_RATIONAL:
                                case TIFF_SRATIONAL:
                                    size = 4;
                                    break;

                                case TIFF_DOUBLE:
                                    size = 8;
                                    break;
                        }

                        array = _TIFFmalloc(wc * size);
                        if (!array) {
                                fprintf(stderr, "No space for %s tag\n",
                                        tagname);
                                return 4;
                        }

                        switch (fip->field_type) {
                            case TIFF_BYTE:
                                for (i = 0; i < wc; i++)
                                    ((uint8 *)array)[i] = atoi(argv[arg_index+i]);
                                break;
                            case TIFF_SHORT:
                                for (i = 0; i < wc; i++)
                                    ((uint16 *)array)[i] = atoi(argv[arg_index+i]);
                                break;
                            case TIFF_SBYTE:
                                for (i = 0; i < wc; i++)
                                    ((int8 *)array)[i] = atoi(argv[arg_index+i]);
                                break;
                            case TIFF_SSHORT:
                                for (i = 0; i < wc; i++)
                                    ((int16 *)array)[i] = atoi(argv[arg_index+i]);
                                break;
                            case TIFF_LONG:
                                for (i = 0; i < wc; i++)
                                    ((uint32 *)array)[i] = atol(argv[arg_index+i]);
                                break;
                            case TIFF_SLONG:
                            case TIFF_IFD:
                                for (i = 0; i < wc; i++)
                                    ((uint32 *)array)[i] = atol(argv[arg_index+i]);
                                break;
                            case TIFF_DOUBLE:
                                for (i = 0; i < wc; i++)
                                    ((double *)array)[i] = atof(argv[arg_index+i]);
                                break;
                            case TIFF_RATIONAL:
                            case TIFF_SRATIONAL:
                            case TIFF_FLOAT:
                                for (i = 0; i < wc; i++)
                                    ((float *)array)[i] = (float)atof(argv[arg_index+i]);
                                break;
                            default:
                                break;
                        }
                
                        if (fip->field_passcount) {
                                ret = TIFFSetField(tiff, fip->field_tag,
                                                   wc, array);
                        } else {
                                ret = TIFFSetField(tiff, fip->field_tag,
                                                   array);
                        }

                        _TIFFfree(array);
                } else {
                        switch (fip->field_type) {
                            case TIFF_BYTE:
                            case TIFF_SHORT:
                            case TIFF_SBYTE:
                            case TIFF_SSHORT:
                                ret = TIFFSetField(tiff, fip->field_tag,
                                                   atoi(argv[arg_index++]));
                                break;
                            case TIFF_LONG:
                            case TIFF_SLONG:
                            case TIFF_IFD:
                                ret = TIFFSetField(tiff, fip->field_tag,
                                                   atol(argv[arg_index++]));
                                break;
                            case TIFF_DOUBLE:
                                ret = TIFFSetField(tiff, fip->field_tag,
                                                   atof(argv[arg_index++]));
                                break;
                            case TIFF_RATIONAL:
                            case TIFF_SRATIONAL:
                            case TIFF_FLOAT:
                                ret = TIFFSetField(tiff, fip->field_tag,
                                                   (float)atof(argv[arg_index++]));
                                break;
                            default:
                                break;
                        }
                }

                if (ret != 1)
                    fprintf(stderr, "Failed to set %s\n", fip->field_name);
                arg_index += wc;
            }
        } else if (strcmp(argv[arg_index],"-sf") == 0 && arg_index < argc-3) {
            FILE    *fp;
            const TIFFFieldInfo *fip;
            char    *text;
            int     len;

            arg_index++;
            fip = GetField(tiff, argv[arg_index]);

            if (!fip)
                return 3;

            if (fip->field_type != TIFF_ASCII) {
                fprintf( stderr,
                         "Only ASCII tags can be set from file. "
                         "%s is not ASCII tag.\n", fip->field_name );
                return 5;
            }

            arg_index++;
            fp = fopen( argv[arg_index], "rt" );
            if(fp == NULL) {
                perror( argv[arg_index] );
                continue;
            }

            text = (char *) malloc(1000000);
            len = fread( text, 1, 999999, fp );
            text[len] = '\0';

            fclose( fp );

            if(TIFFSetField( tiff, fip->field_tag, text ) != 1) {
                fprintf(stderr, "Failed to set %s from file %s\n", 
                        fip->field_name, argv[arg_index]);
            }

            _TIFFfree( text );
            arg_index++;
        } else {
            fprintf(stderr, "Unrecognised option: %s\n",
                    argv[arg_index]);
            usage();
        }
    }

    TIFFRewriteDirectory(tiff);
    TIFFClose(tiff);
    return 0;
}

/* vim: set ts=8 sts=8 sw=8 noet: */
